subsys/net/lib/conn_mgr/conn_mgr.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(conn_mgr, CONFIG_NET_CONNECTION_MANAGER_LOG_LEVEL);
+
+#include <init.h>
+#include <kernel.h>
+#include <errno.h>
+#include <net/net_core.h>
+#include <net/net_if.h>
+#include <net/net_mgmt.h>
+
+#include <conn_mgr.h>
+
+u16_t iface_states[CONN_MGR_IFACE_MAX];
+
+K_SEM_DEFINE(conn_mgr_lock, 1, UINT_MAX);
+
+static enum net_conn_mgr_state conn_mgr_iface_status(int iface_idx)
+{
+	if (iface_states[iface_idx] & NET_STATE_IFACE_UP) {
+		return NET_CONN_MGR_STATE_CONNECTED;
+	}
+
+	return NET_CONN_MGR_STATE_DISCONNECTED;
+}
+
+#if defined(CONFIG_NET_IPV6)
+static enum net_conn_mgr_state conn_mgr_ipv6_status(int iface_idx)
+{
+	if ((iface_states[iface_idx] & CONN_MGR_IPV6_STATUS_MASK) ==
+	    CONN_MGR_IPV6_STATUS_MASK) {
+		NET_DBG("IPv6 connected on iface index %u", iface_idx);
+		return NET_CONN_MGR_STATE_CONNECTED;
+	}
+
+	return NET_CONN_MGR_STATE_DISCONNECTED;
+}
+#else
+#define conn_mgr_ipv6_status(...) NET_CONN_MGR_STATE_CONNECTED
+#endif /* CONFIG_NET_IPV6 */
+
+#if defined(CONFIG_NET_IPV4)
+static enum net_conn_mgr_state conn_mgr_ipv4_status(int iface_idx)
+{
+	if ((iface_states[iface_idx] & CONN_MGR_IPV4_STATUS_MASK) ==
+	    CONN_MGR_IPV4_STATUS_MASK) {
+		NET_DBG("IPv4 connected on iface index %u", iface_idx);
+		return NET_CONN_MGR_STATE_CONNECTED;
+	}
+
+	return NET_CONN_MGR_STATE_DISCONNECTED;
+}
+#else
+#define conn_mgr_ipv4_status(...) NET_CONN_MGR_STATE_CONNECTED
+#endif /* CONFIG_NET_IPV4 */
+
+static void conn_mgr_notify_status(int iface_idx)
+{
+	struct net_if *iface = net_if_get_by_index(iface_idx + 1);
+
+	if (iface_states[iface_idx] & NET_STATE_CONNECTED) {
+		NET_DBG("Iface %p connected", iface);
+		net_mgmt_event_notify(NET_EVENT_L4_CONNECTED, iface);
+	} else {
+		NET_DBG("Iface %p disconnected", iface);
+		net_mgmt_event_notify(NET_EVENT_L4_DISCONNECTED, iface);
+	}
+}
+
+static void conn_mgr_act_on_changes(void)
+{
+	int idx;
+
+	for (idx = 0; idx < CONN_MGR_IFACE_MAX; idx++) {
```

Comment:
Actually I meant ARRAY_SIZE() macro here.

Reply:
it is the same exact variable used for the array size and here, why would that differ at some point?

Reply:
If someone sets `u16_t iface_states[CONN_MGR_IFACE_MAX + 1];` then this for loop is wrong. If we use ARRAY_SIZE() here, it is always correct regardless how iface_states is set.

Reply:
ok if you insist (though it would be a bug to raise the size of the array like that. It would be as buggy as doing array_size() <-/+> number in the for loop)